balancer/leastrequest/leastrequest.go

```diff
@@ -162,18 +162,18 @@ func (p *picker) Pick(balancer.PickInfo) (balancer.PickResult, error) {
 			pickedSC = &sc
 			continue
 		}
-		if *sc.numRPCs < *pickedSC.numRPCs {
+		if sc.numRPCs.Load() < pickedSC.numRPCs.Load() {
```

Comment:
Q for @zasweq: should we cache `pickedSC.numRPCs`, or reload it on every iteration as we are currently doing?

```go
for ... {
	index := ...
	sc := ...
	scRPCs := sc.numRPCs.Load()
	if pickedSC == nil || scRPCs < pickedSCRPCs{
		pickedSC = &sc
		pickedSCRPCs = scRPCs
	}
}
```

Reply:
I see what you're saying. I think both caching it and the way it is fall within the language described here: https://github.com/grpc/proposal/blob/master/A48-xds-least-request-lb-policy.md#outstanding-request-counter-raciness. Caching it before the iteration would get rid of the race where other threads and picks can add or subtract between iterations. "While reading the outstanding request counters of samples in the picker, previously read values may become outdated" - specifically, this language I think encompasses a. scs that have already been checked and b. the picked sc. As Eric mentioned offline, I think eventually this algorithm goes towards consistency even if the checks aren't perfect.

Reply:
Actually, just talked to Eric about this and he agrees with your comment. a. loads are expensive, particularly if you get a subsequent write in another Pick call, since it requires exclusive access similar to a rw lock. Thus, it is more performant to perform the operation once. b. Wrt atomics, Eric has the philosophy of only Loading once, as it makes the code easier to reason about and also prevents concurrent operations on that data breaking assumptions and logic about the code. After this is merged I'll change it and write a test for concurrent RPC's.